File: javascript/linked-list/linked-list.js
export class LinkedList {

    constructor() {
        this.size = 0;
        this.head = null;
        this.tail = null;
    }

    push(data) {

        let node = new Node(data);
        node.previous = this.tail;

        if (this.size === 0) {
            this.head = node;
        } else {
            this.tail.next = node;
        }

        this.tail = node;
        this.size++;
    }

    pop() {

        let data = this.tail.value;
        this.size--;

        if (this.size === 0) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.previous;
            this.tail.next = null;
        }

        return data;
    }

    shift() {
        let data = this.head.value;
        this.size--;

        if (this.size === 0) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
            this.head.prev = null;
        }

        return data;
    }

    unshift(data) {

        let node = new Node(data);
        node.next = this.head;

        if (this.size === 0) {
            this.head = node;
            this.tail = node;
        } else {
            this.head.prev = node;
            this.head = node;
        }

        this.size++;
    }

    delete(item) {
        let nodeToDelete = null;
        let currentNode = this.head;

        while (currentNode !== null) {
            if (currentNode.value === item) {
                if (currentNode.previous !== null) {
                    currentNode.previous.next = currentNode.next;
                } else {
                    this.head = currentNode.next;
                }

                if (currentNode.next !== null) {
                    currentNode.next.previous = currentNode.previous;
                } else {
                    this.tail = currentNode.previous;
                }

                this.size--;
                break;
            }

            currentNode = currentNode.next;
        }
    }

    count() {
        return this.size;
    }
}

class Node {

    constructor (value) {
        this.value = value;
        this.previous = null;
        this.next = null;
    }
}

File: python/linked-list/linked_list.py
class Node:
    def __init__(self, value, succeeding=None, previous=None):
        self.value = value
        self.succeding = succeeding
        self.previous = previous


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0


    def push(self, value):
        new_node = Node(value)

        if self.size == 0:
            self.head = new_node
        else:
            new_node.previous = self.tail
            self.tail.succeeding = new_node

        self.tail = new_node
        self.size += 1

    def pop(self):
        value = self.tail.value

        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.previous
            self.tail.succeeding = None

        self.size -= 1
        return value

    def shift(self):
        value = self.head.value

        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.succeeding
            self.head.previous = None

        self.size -= 1
        return value

    def unshift(self, value):
        new_node = Node(value)

        if self.size == 0:
            self.tail = new_node
        else:
            new_node.succeeding = self.head
            self.head.previous = new_node

        self.head = new_node
        self.size += 1

